Workshop3 amendments

janvanmansum opened the pull request:
Review and amend
- [x] higher order functions
- [x] functional operators
- [x] option
- [x] option-flatmap
- [x] list-for-comprehension
- [x] try
- [x] other collections
- [x] assigments

janvanmansum committed:
Working on amendments

janvanmansum committed:
Made link paths relative

janvanmansum committed:
WIP

janvanmansum committed:
WIP

janvanmansum committed:
WIP

janvanmansum committed:
WIP

janvanmansum committed:
WIP

janvanmansum commented:
@rvanheest don't merge yet! I am still working on it.

rvanheest commented:
Same here, added some extra things to the last assignment this morning that I still need to write tests for... Good luck and have a nice weekend :wink:

janvanmansum committed:
WIP

janvanmansum commented:
Some notes (I'll be expanding this comment as I review more text):
- First of all, good job! I like the structure of the workshop and how you build things up. That said ...
- Please, make the markdown also readable as plain text. This is kind of the idea with markdown: the resulting html is nicely formatted, but the "source code" should also be fit for human consumption. In particular:
  - Use the underlining style for the h1 and h2 headings rather than the `#`-style
  - Use lines of reasonable lenght, say, 160 or so.
  - Use two empty lines between sections, one between paragraphs
  - If links are very long, define them below the paragraph in the `[link name]: <url>` style
  - Use the back ticks consistently when refering to Scala-names, such as `String` and `null`, but leave them out if you use words like "string" in a more abstract sense.
- For writing markdown in IntelliJ, use the (free version of) "Markdown navigator" plugin, and also install the "gfm" plugin. The first has great syntax highlighting (put it on GitHub-mode in Preferences), the second can be configured to show an extra preview tab (next to the Navigator one). The thing is that the free version of Navigator does not display tables, whereas gfm does.
- The bulk of my changes are formatting and some minor language or style errors. The disadvantage of doing such a large "amendment" pull request if of course that it become rather difficult to see where I _did_ change the content, so I'll try to list those instances in the following points.
- I have made some changes to the `fold`-examples. I found them very confusing, so I think I have improved them somewhat, but maybe I got it all wrong, so please review my changes.
- The examples `makeStringOfLength` may throw the reader off in some ways:
  - what it does may be a bit hard to see at first glance. Stating more explicitly what it does may help; also making clear that it is not really important what it does.
  - you go through several iterations and call the `makeStringOfLength2` etc, in which `2` might be mistaken for the lenght of the string produced.
- In quite a number of places I have replaced lambda's like this `x => x + 1` with `_ + 1`. I think we should really not encourage a long-winded style, here. Also, I have undone unnecessary wrapping of `println` where I encountered it.
- I have converted most references to API-functions to links into the Scaladocs.

janvanmansum committed:
WIP

rvanheest commented:
update this PR with the latest commits. added some extra (small) things.

janvanmansum committed:
WIP

janvanmansum commented:
I have incorporated your latest changes (at least I hope I got all of them). Please, merge my changes first, before making any further changes yourself. Otherwise conflicts will prove too difficult to resolve.

rvanheest commented:
I can't merge yet...
Let's look at it tomorrow.

janvanmansum commented:
Okay

rvanheest commented on .gitignore:
```diff
@@ -2,3 +2,5 @@
 
 *.iml
+
+/*.sc
```
Not a fan of ignoring `*.sc` in this case. It is used in workshop 1, 1a and 2 quite frequently. We either rewrite them to *.md files (e.g. make another PR for this in the (near) future) or we remove the *.sc here.

rvanheest commented on src/main/scala/workshop3/README.md:
```diff
@@ -1,17 +1,24 @@
-* [for-comprehension on `List`](list-for-comprehensions.md)
-* [`Try`](try.md) - *introducting the `Try` type as **the** way of handling errors in a functional way*
-* [other collections](other-collections.md) - *conversion between Java and Scala collections and some more operators*
+1. [higher-order functions](01%20higher-order-functions.md) - *getting to know what a higher-order function is, how it looks and how you use it*
```
I like the numbering of the files and the way of listing them here!

rvanheest commented on src/main/scala/workshop3/01 higher-order-functions.md:
````diff
@@ -0,0 +1,180 @@
+numberOfBandC("abba")
+```
+
+Looking back at this code makes you realize that your are repeating yourself a lot (in violation of the [DRY]-principle). 
````
Great place to mention DRY indeed!

rvanheest commented on src/main/scala/workshop3/01 higher-order-functions.md:
```diff
@@ -0,0 +1,180 @@
+[DRY]: https://en.wikipedia.org/wiki/Don%27t_repeat_yourself
+
+A higher-order function is a function that takes other functions as arguments or returns a function as result. This is different 
+from the functions above that just take 'normal' values as their arguments. A Higher-order function may look like this:
```
`A Higher-order ...` --> `A higher-order`

rvanheest commented on src/main/scala/workshop3/01 higher-order-functions.md:
```diff
@@ -0,0 +1,180 @@
+* `X` is the return type of the higher-order function
+
+Note that higher-order functions can also take *'normal'* values as their arguments. The term 'higher-order' 
+just signifies that one or more arguments (or the return value) are functions.
```
Didn't think of the return value indeed. Thanks for adding this!

rvanheest commented on src/main/scala/workshop3/01 higher-order-functions.md:
````diff
@@ -0,0 +1,180 @@
+```
+
+Basically, you can think of these functions as the same functions as the ones before, but without 
+a name (hence: "anonymous")
````
add a `.` at the end

rvanheest commented on src/main/scala/workshop3/02 functional-operators.md:
```diff
@@ -0,0 +1,237 @@
+
+The aggregator function is first applied to the seed value and the first (`foldLeft`) or last (`foldRight`) element of the list, then
+the result of that is applied to the next value, etc. The final result is the result of the last application of the aggregator.
+The following pictures, taken from Wikipedia depict this process. Not the execution in both cases proceeds from the bottom of the
```
add a `,` after `taken from Wikipedia`

rvanheest replied:
"Note that execution in both cases proceeds from the bottom to the top."

rvanheest commented on src/main/scala/workshop3/02 functional-operators.md:
```diff
@@ -0,0 +1,237 @@
+
+The aggregator function is first applied to the seed value and the first (`foldLeft`) or last (`foldRight`) element of the list, then
+the result of that is applied to the next value, etc. The final result is the result of the last application of the aggregator.
+The following pictures, taken from Wikipedia depict this process. Not the execution in both cases proceeds from the bottom of the
```
I don't really understand the last sentence of this paragraph.

rvanheest commented on src/main/scala/workshop3/02 functional-operators.md:
````diff
@@ -0,0 +1,237 @@
+List('a', 'b', 'c', 'd').foldRight(0)((_, count) => count + 1)
+```
+
+Notice that the result of subtraction is the same for `foldLeft` and `foldRight`, but execution follows a different path
````
This paragraph about folding with subtraction was intended differently. For all the other examples above it doesn't matter whether you use `foldLeft` or `foldRight` in combination with `_ operator _`. However, `List(1,2,3,4).foldLeft(20)(_ - _) == 10` whereas `List(1,2,3,4).foldRight(20)(_ - _) == 18`. This is because subtraction is not associative. I see I made a mistake myself in my version as well by reversing `rest` and `i` in `val res = ...`. Could you revert your changes and look at it again with this explanation?

rvanheest commented on src/main/scala/workshop3/02 functional-operators.md:
```diff
@@ -0,0 +1,237 @@
+(Note on the shorter version: when using multiple underscores to access function parameters without naming them, each underscore represents 
+the next parameter in order, that is why you can only use them once!)
+
+(Note that you are now more limited in the aggregator functions you can use!)
```
What do you mean with this note?

rvanheest commented on src/main/scala/workshop3/03 option.md:
````diff
@@ -0,0 +1,122 @@
+With this we can refactor the function above to return an `Option[String]` instead.
+
+```scala
+def makeStringOfLength2(n: Int): Option[String] = {
````
You are right about the ambiguity here. Didn't notice it before. Would you suggest removing the numbers and calling all iterations the same?

rvanheest commented on src/main/scala/workshop3/04 option-flatmap.md:
```diff
@@ -33,42 +37,54 @@ Just as before we can rewrite this such that `getBar` and `getBaz` return an `Op
+What we end up with, however, is not what we wanted: `Option[Option[String]`. Now we got two `Option`s nested 
+before we have the `String`. Both `Option`s denote the fact that the value inside could potentially be a `null`.
+Of course we can do with only one `Option` and therefore we can flatten the `Option[Option[String]]` 
+into an `Option[String]` by using `flatMap` instead of `map`. This operator first maps a value inside an `Option` 
```
add a link to `flatMap` in the Scala API, as you did with all others.

rvanheest commented on src/main/scala/workshop3/README.md:
```diff
@@ -1,17 +1,24 @@
-* [Scala Monads](https://youtu.be/Mw_Jnn_Y5iA) - *A good starter explanation of the `Option`, `Try` and `List` monads <sup>that shows remarkable similarities with our explanation :stuck_out_tongue_winking_eye:...</sup>*
+Further reading/watching
+------------------------
+
```
I've added an extra link here. Probably that is why you can't merge...

rvanheest commented on src/main/scala/workshop3/assignments/FileIO.md:
```diff
@@ -1,29 +1,61 @@
 * `report2.txt` will contain text formatted as follows: `<customer_name> has to pay <total_price>`
 
-To read from file and write to file, we use the *Scala-ARM* library to do automatic resource management. Since this library defines `map` and `flatMap` on its resources, you can use this in a for-comprehension (which you are going to do later in this assignment).
+To read from file and write to file, we use the *Scala-ARM* library to do automatic resource management. Since this l
```
`l`\n`ibrary`...

rvanheest commented:
Thanks for the review. Many useful additions! Added a few comments that I would like you to look at before we can merge.
